src/dmd/root/serializer.d

```diff
@@ -0,0 +1,217 @@
+     *
+     * Params:
+     *  value = the value to serialize
+     *  isTopLevel = indicates if this is the first value to be serialized
+     */
+    void serializeBasicType(T)(T value, bool isTopLevel)
+    if (isBasicType!T)
+    {
+        previousType = Type.basic;
+
+        if (isTopLevel)
+            append(' ');
+
+        append(value);
+    }
+
+    /**
+     * Increases the indentation level for the duration of the given block.
+     *
+     * Params:
+     *  block = the code to execute while the indentation level has been
+     *      increased
+     */
+    void indent(void delegate() block)
+    {
+        level++;
+        scope (exit) level--;
+        block();
+    }
+
+    /// Writes out the current level of indentation to the buffer.
+    void writeIndentation()
+    {
+        if (shouldIdent)
+        {
+            foreach (_ ; 0 .. level * indentation)
+                buffer.writeByte(' ');
+        }
+    }
+
+    /// Adds a newline to the buffer.
+    void newline()
+    {
+        buffer.writenl();
+        shouldIdent = true;
+    }
+
+    /**
+     * Appends the given string(s) to the buffer.
+     *
+     * Params:
+     *  strings = the string(s) to append to the buffer
+     */
+    void append(string[] strings ...)
+    {
+        writeIndentation();
+
+        foreach (str ; strings)
+            buffer.writestring(str);
+
+        shouldIdent = false;
+    }
+
+    /**
+     * Appends the given value to the buffer.
+     *
+     * Params:
+     *  value = the value to append to the buffer
+     */
+    void append(T)(T value)
+    {
+        alias U = Unqual!T;
+
+        static const(char)* printfFormatter(T)()
+        {
+            static if (is(U == byte) || is(U == ubyte))
+                return "%d";
+            else static if (is(U == char))
+                return "%c";
+            else static if (is(U == short) || is (U == int))
+                return "%d";
+            else static if (is(U == ushort) || is (U == uint))
+                return "%u";
+            else static if (is(U == long))
+                return "%lld";
+            else static if (is(U == ulong))
+                return "%llu";
+            else static if (is(U == float) || is(U == double))
+                return "%g";
+            else
+                static assert(false, "Serializing a value of type `" ~ U.stringof ~ "` is not supported");
+        }
+
+        writeIndentation();
+
+        static if (is(U == bool))
+            buffer.writestring(value ? "true" : "false");
+        else static if (is(U == char) || is(U == wchar) || is(U == dchar))
+            buffer.writeUTF8(value);
+        else static if (is(U == real))
+            buffer.write(value);
+        else
+            buffer.printf(printfFormatter!U, value);
+
+        shouldIdent = false;
+    }
+}
+
+private:
+
+/// Evaluates to `true` if `T` is a basic type, otherwise `false`.
+template isBasicType(T)
```

Comment:
This duplicates `isTypeBasic()`

Reply:
`isTypeBasic` is a member of the class `Type`. This is intended to pass in native D types, like `int` and `char[]`. It's not what the compiler views as a basic type. It's used by the serializer to determine if the value needs to be further decomposed or if it can be transformed into a textual representation as is.